Ticket#7113 Validación de DOI externo al momento de registrar DOI

FacundoAdorno opened the pull request:
Se agregó una validación de la existencia de un DOI externo (principalmente guardado en el metadato 'sedici.identifier.other') al momento de ejecutar el método _#register()_ o _#registerOnline()_ en el DOIIdentifierProvider.java.

De esta manera, se evita que se creen tuplas innecesarias en la tabla 'doi' de items en la espera de registrarse (TO_BE_REGISTERED) cuando en realidad no es necesaria su registración por la existencia de otro identificador DOI externo al repositorio.

FacundoAdorno committed:
-#7113 Se agregó una validación de existencia de DOI externo

Esto para evitar que se creen tuplas en la tabla 'doi' de items que
estén esperando a registrarse (TO_BE_REGISTERED) en la RA cuando en
realidad ya tienen un DOI asignado de antemano (y diferente al utilizado
en el repositorio).

FacundoAdorno committed:
-#7113 Validación de DOI externo al momento de registrar online

Ya que si acá no se verifica se puede crear una tupla innecesaria en la
tabla 'doi' de un ítem que tiene un DOI externo. El método
registerOnline se utiliza principalmente cuando se utiliza el
comando "./dspace doi-organiser -r" (o similares) desde la línea de
comandos.

santit96 commented:
Estuve probando y al parecer anda todo bien, al momento de hacer el install de un item el log que se muestra desde el DOIIdentifierProvider cuando se intenta meter al nuevo item en la tabla de DOI pero ya tiene un doi en sedici.identifier.other es este:

`2021-09-15 13:02:12,115 INFO  org.dspace.identifier.DOIIdentifierProvider @ Couldn't register doi for DSO with handle 10915/119304, the DSO already has an external DOI.
`

santit96 commented:
Y si por alguna razón está encolado para ser registrado un ítem que ya tiene un DOI asignado, lo que se nos muestra al momento de querer registrarlo desde la linea de comandos es:

`It wasn't possible to register this identifier: doi:10.35537/10915/119303
org.dspace.identifier.doi.DOIIdentifierException: Cannot register for new DOI when item already has an external DOI.
	at org.dspace.identifier.DOIIdentifierProvider.registerOnline(DOIIdentifierProvider.java:320)
	at org.dspace.identifier.doi.DOIOrganiser.register(DOIOrganiser.java:508)
	at org.dspace.identifier.doi.DOIOrganiser.runCLI(DOIOrganiser.java:238)
	at org.dspace.identifier.doi.DOIOrganiser.main(DOIOrganiser.java:83)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:226)
	at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:78)
`

Esto es parecido al mensaje que se mostraba antes de los cambios pero es un poco más preciso

santit96 commented:
Funciona todo bien, por mi mergear

arieljlira commented on dspace/modules/additions/src/main/java/ar/edu/unlp/sedici/dspace/identifier/doi/service/DOIFilterServiceImpl.java:
```diff
@@ -13,14 +13,45 @@
+        if (metadataList.length > 0) {
+            for (Metadatum md : metadataList) {
+                try {
+                    String doi = DOI.formatIdentifier(md.value);
```
Analizar si se puede evitar el format en ids que no son dois

FacundoAdorno committed:
-#7113 Agrego una validacion antes de formatear un posible DOI y así corroborar si es un DOI

Esto es para evitar hacer este formateo en todos los los metadatos
sedici.identifier.other.

FacundoAdorno commented:
Santi, ahi agregue la validacion del substring 'doi'. Si lo podes cuando puedas probar una vez mas buenisimo y lo mergeamos

santit96 commented:
> Santi, ahi agregue la validacion del substring 'doi'. Si lo podes cuando puedas probar una vez mas buenisimo y lo mergeamos

Probé con estos nuevos cambios y sigue funcionando bien. No agrega a la cola de registro a los items que ya tienen doi en sedici.identifier.other y a los que ya estan en la cola y tienen doi en ese metadato tampoco los registra
